public/extensions/breakout.js

```diff
@@ -0,0 +1,116 @@
+// @ts-check
+
+import daily from "./core.js";
+import Socket from "./socketiostate.js";
+
+let call;
+let socket;
+
+// Initial state
+let state = {
+  initialRoomUrl: "",
+  breakoutStarted: false,
+  participants: [{ user_name: "", roomUrl: "" }],
+};
+
+daily.beforeCreateFrame((parentEl, properties) => {
+  if (!properties.customTrayButtons) {
+    properties.customTrayButtons = {};
+  }
+
+  properties.customTrayButtons = {
+    ...properties.customTrayButtons,
+    toggleBreakout: {
+      iconPath: "https://www.svgrepo.com/show/207394/flash.svg",
+      label: "Breakout",
+      tooltip: "Breakout",
+    },
+  };
+
+  return [parentEl, properties];
+});
+
+daily.afterCreateFrame(async (c) => {
+  call = c;
+  call.on("custom-button-click", (e) => {
+    if (e.button_id !== "toggleBreakout") {
+      return;
+    }
+
+    state.breakoutStarted ? end() : start();
+  });
+});
+
+async function handleOnStateUpdate(s = {}) {
+  state = { ...state, ...s };
+
+  const localParticipant = call.participants().local;
+
+  const roomUrl = state.participants.find(
+    (p) => p.user_name === localParticipant?.user_name
+  )?.roomUrl;
+
+  if (roomUrl) {
+    try {
+      await call.leave();
+      await call.join({ url: roomUrl });
+    } catch (err) {
+      console.error("Failed to join new room:", err);
+    }
+  }
+}
+
+export function connect({ room = "", domain = "" }) {
```

Comment:
Since we're just connecting to a socket based on the domain and room name, we _could_ remove the need to `connect` from the top-level app page by doing this in `beforeCreateFrame`. I think it's still a good idea to have it as a private method, though.